content/blog/2017/2017-01-31-pipeline-scalability-best-practice.adoc

```diff
@@ -0,0 +1,197 @@
+    ..  While normal pipeline is restricted to serializable local variables
+(see appendix at bottom), @NonCPS functions can use more complex,
+nonserializable types internally (ex regex matchers, etc). Parameters
+and return types should still be Serializable, however.
+    ... *Gotcha:* improper usages are not guaranteed to raise an error with
+normal pipeline (optimizations may mask the issue), but it is unsafe to
+rely on this behavior.
+.  *Prefer external scripts/tools for complex or CPU-expensive
+processing* rather than Groovy language features. This offloads work
+from the master to external executors, allowing for easy scale-out of
+hardware resources. It is also generally easier to test because these
+components can be tested in isolation without the full on-master
+execution environment.
+    ..  Many software vendors will provide easy command-line clients for
+their tools in various programming languages. These are often robust,
+performant, and easy to use. Plugins offer another option (see below).
+    ..  Shell or batch steps are often the easiest way to integrate these
+tools, which can be written in any language. For example: sh “java -jar
+client.jar $endPointUrl $inputData” for a Java client, or sh “python
+jiraClient.py $issueId $someParam” for a python client. +
+    ..  Examples of places to use this:
+        ...  XML/JSON parsing and transformation (ex: curl + xmllint for XML
+xpath, or curl + jq for JSON)
+        ...  Templating large files
+        ...  Nontrivial integration with external APIs (for bigger vendors,
+consider a Jenkins plugin if a quality offering exists)
+        ...  Simulations/complex calculations
+        ...  Business logic
+.  *Consider existing plugins for external integrations.* Jenkins has a
+wealth of plugins, especially for source control, artifact management,
+deployment systems, and systems automation. These can greatly reduce the
+amount of pipeline code to maintain. Well-written plugins may be be
+faster and more robust than pipeline equivalents.
+    ..  Consider both plugins and command-line clients (above) -- one may be
+easier than the other.
+    ..  Plugins may be of widely varying quality. Vendor-maintained plugins
+with a large install base are often of good quality. Third-party plugins
+with limited installations may actually be worse than writing a little
+custom pipeline code.
+    ..  As a last resort, if there is a good-quality plugin that is not
+pipeline-enabled, it is fairly easy to write a pipeline wrapper to
+integrate it or write a custom step that will invoke it.
+.  *Assume things will go wrong:* don’t rely on workspaces being clean
+of the remnants from previous executions, clean explicitly where needed.
+Make use of timeouts and retry steps (that’s what they’re there for).
+    ..  Within a git repository, ‘git clean -fdx’ is a good way to
+accomplish this and reduces the amount of SCM cloning
+.  *DO use parameterized pipelines and variables to make your pipeline
```

Comment:
Ehhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhhh. Personally I think this advice encourages more crazy parameterized freestyle-job-chaining patterns. IMHO Multibranch addresses where many people had previously used parameterized projects.